File: lab-4-follow-the-gap-team-2-1-main/gap_follow/src/reactive_node.cpp
#include "rclcpp/rclcpp.hpp"
#include <string>
#include <vector>
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"


// vehicle parameters, for bubble radius.
const float CHASSIE_WIDTH = 0.2032;
const float TIRE_WIDTH = 0.0381;
const float CAR_WIDTH = CHASSIE_WIDTH + 2 * TIRE_WIDTH;
const float MAX_STEER = 0.36;


class ReactiveFollowGap : public rclcpp::Node {

public:
    ReactiveFollowGap() : Node("reactive_node")
    {
        // 1. LiDAR subscriber
        subscriber_scan = this -> create_subscription<sensor_msgs::msg::LaserScan>(
            "scan", 10, std::bind(&ReactiveFollowGap::lidar_callback, this, std::placeholders::_1)
        );

        // 2. Car actuator publisher
        publisher_ = this -> create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(
            "drive", 10
        );
        
        // 3. Declare parameters
        // param for preprocessing
        this->declare_parameter("reject_thres", 2.5);           // preprocess depth 
        this->declare_parameter("window_radius", 2);            // avg over window 
        // param for finding max gap
        this->declare_parameter("consecutive_hits", 5);         // # of consecutive hits at depth > distance_thres count as a valid gap.
        this->declare_parameter("distance_thres", 2.0);         // threshold depth of valid gap
        // param for bubble
        this->declare_parameter("bubble_radius_factor", 0.7);   // bubble radius
        this->declare_parameter("disparities_thres", 0.2);      // disparity extension threshold for depth difference.
        // param for steering PD control
        this->declare_parameter("k_p", 1.0);                    // Steering Gains
        this->declare_parameter("k_d", 0.1);
        this->declare_parameter("k_i", 0.0001);
        this->declare_parameter("alpha", 0.4);                  // steering smoothing
    }

private:
    std::string lidarscan_topic = "/scan";
    std::string drive_topic = "/drive";

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscriber_scan;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr publisher_;

    // laser scan params
    int range_length;       // length of ranges
    float range_min;        // valid range
    float range_max;
    float angle_min;        // angle range
    float angle_max;
    float angle_increment;
    int front_idx_start;    // idx of scans in -90~90 degrees ahead
    int front_idx_end;

    // PD controller history
    float integral      = 0.0;
    float prev_angle    = 0.0;
    float prev_steer    = 0.0; 
    
    int angle_to_index(float angle) {
        // Helper function, find index in ranges given angle.
        if (angle < angle_min) {
            return 0;
        }
        else {
            return (int)round((std::min(angle, angle_max) - angle_min) / angle_increment);
        }

    }

    float index_to_angle(int index) {
        // Helper function, find angle given ray index.
        float angle = angle_min + index * angle_increment;
        return angle;
    }

    void preprocess_lidar(const float* ranges, float* processed_ranges)
    {   
        // Preprocess the LiDAR scan array. Expert implementation includes:
        // 1.Setting each value to the mean over some window
        // 2.Rejecting high values (eg. > 3m)
        int window_radius = this->get_parameter("window_radius").as_int();
        float reject_thres = this->get_parameter("reject_thres").as_double();
        for (int i=0; i < range_length; i++) {
            // Find avg in window
            float avg = 0.0;
            float count = 0;
            for (int j=-window_radius; j<=window_radius; j++) {
                if (i+j>=0 && i+j<range_length) {
                    avg += ranges[i+j]; 
                    count++;
                }
            }
            avg = avg / count;
            if (avg > reject_thres) {
                processed_ranges[i] = reject_thres;
            }
            else {
                processed_ranges[i] = avg;
            }
        }
        return;
    }

    void find_max_gap(const float* ranges, int* indices)
    {   
        // Return the start index & end index of the max gap in free_space_ranges
        int consecutive_hits = this->get_parameter("consecutive_hits").as_int();
        float distance_thres = this->get_parameter("distance_thres").as_double();
        int gap_counter     = 0; //count the # of consective gaps
        int gap_max_size    = 0; //save the max-gap
        int gap_start_idx   = 0; //save the start index
        for (int i=0; i<range_length; i++){
            if (ranges[i] >= distance_thres){
                gap_counter += 1;
                if (gap_counter == 1) {
                    gap_start_idx = i;
                }
                else {
                    if (gap_counter >= consecutive_hits && gap_counter > gap_max_size){
                        gap_max_size = gap_counter;
                        indices[0] = gap_start_idx;
                        indices[1] = i;
                    }
                }
            }
            else {
                gap_counter = 0;
            }
        }
        // RCLCPP_INFO(this->get_logger(), "gap_max: %f", gap_max);
        return;
    }

    int find_best_point(float* ranges, const int* gap_indices) {   
        // TODO: If ranges has consecutive max values, use the one in the middle.
        // Start_i & end_i are start and end indicies of max-gap range, respectively
        // Return index of best point in ranges
	    // Naive: Choose the furthest point within ranges and go there
        int index;
        float farthest_distance = ranges[gap_indices[0]];
        int consecutive_count = 1;
        int fartheset_index_start = gap_indices[0];
        // *indice = gap_start_idx;
        for (int i = gap_indices[0] + 1; i <= gap_indices[1]; i++) {
            if (ranges[i] > farthest_distance) {
                farthest_distance = ranges[i];
                fartheset_index_start = i;
                consecutive_count = 1;
            } 
            else if (ranges[i] == farthest_distance) {
                consecutive_count++;
            }
        }
        index = fartheset_index_start + consecutive_count / 2;
        return index;
    }

    // void find_disparities(const float* ranges, std::vector<int>& bubble_indices) {
        
    //     return;
    // }

    std::vector<std::pair<int, int>> find_bubbles(const float* ranges) {
        // Find center of bubbles
        float disparities_thres = this->get_parameter("disparities_thres").as_double();
        std::vector<std::pair<int, int>> bubbles_indices; // (index, filling pattern)
        float closest_distance = 1000.0;
        int closest_idx;
        for (int i=0; i < range_length; i++) {
            if (ranges[i] < closest_distance) {
                closest_idx = i;
                closest_distance = ranges[i];
            }
            if (ranges[i] - ranges[i-1] > disparities_thres) {
                if (i > 0) {
                    bubbles_indices.push_back(std::make_pair(i-1, 1));
                }
            }
            else if (ranges[i]-ranges[i-1]<-disparities_thres) {
                bubbles_indices.push_back(std::make_pair(i, -1));
            }
            
        }
        bubbles_indices.push_back(std::make_pair(closest_idx, 0));
        return bubbles_indices;
    }

    void create_bubbles(float* ranges, std::vector<std::pair<int, int>>& bubble_indices) {
        // create bubbles given a set of bubble centers.
        // scans that fall into the bubble will become 0.
        float bubble_radius_factor = this->get_parameter("bubble_radius_factor").as_double();
        float bubble_radius = bubble_radius_factor * CAR_WIDTH;
        float center_angle;
        float center_distance;
        float fill_value;
        int center_index;
        int direction_flag;
        float theta; // angle between center line & tangent line of the bubble
        int bubble_index_start;
        int bubble_index_end;
        std::vector<std::tuple<int, int, float>> zero_indices;
        for (const std::pair<int, int>& p : bubble_indices) {
            center_index = p.first;
            direction_flag = p.second;
            center_angle = index_to_angle(center_index);
            center_distance = ranges[center_index];
            if (center_distance > bubble_radius) {
                theta = abs(asin(bubble_radius / center_distance));
            }
            else {
                theta = M_1_PI / 12;
                std::cout << center_index << "INSIDE BUBBLE "  << center_distance << std::endl;
            }
            if (direction_flag == 1) {
                bubble_index_start = center_index;
                bubble_index_end =   angle_to_index(center_angle+theta);
            } 
            else if (direction_flag == -1) {
                bubble_index_start = angle_to_index(center_angle-theta);
                bubble_index_end =   center_index;
            } 
            else {
                bubble_index_start = angle_to_index(center_angle-theta);
                bubble_index_end =   angle_to_index(center_angle+theta);
            }            
            
            zero_indices.push_back(std::make_tuple(bubble_index_start, bubble_index_end, center_distance));
            std::cout << "theta " << theta << " center_dis " << center_distance << 
            " BUBBLE INDEX START" << bubble_index_start << ", " <<  bubble_index_end << std::endl;

        }
        for (const auto& p : zero_indices) {
            bubble_index_start = std::get<0>(p);
            bubble_index_end = std::get<1>(p);
            fill_value = std::get<2>(p);
            for (int i=bubble_index_start; i <= bubble_index_end; i++) {
                ranges[i] = fill_value;
            }
        }
        
        return;
    }

    float get_velocity(float distance) {
        // TODO: Tweak 2: Choose velocity based on gap distance
        float low_vel = 0.5;
        float high_vel = 3.0;
        float atten = 1.5;
        float velocity = (high_vel - low_vel) * exp(-abs(distance) * atten) + low_vel;
        return velocity;
    }

        // """ Get desired velocity based on current error
        // """
        // # Speed is exponential w.r.t error
        // low_vel = self.get_parameter('low_vel').get_parameter_value().double_value
        // high_vel = self.get_parameter('high_vel').get_parameter_value().double_value
        // atten = self.get_parameter('velocity_attenuation').get_parameter_value().double_value

        // return (high_vel - low_vel) * np.exp(-abs(error) * atten) + low_vel


    float get_steering_angle(float point_angle) {
        // TODO: steering angle control
        float k_p = this->get_parameter("k_p").as_double();
        float k_d = this->get_parameter("k_d").as_double();
        float alpha = this->get_parameter("alpha").as_double();
        float steer_angle = k_p * point_angle + k_d * (point_angle - prev_angle);
        steer_angle = std::max(-MAX_STEER, std::min(steer_angle, MAX_STEER));
        prev_angle = point_angle;
        steer_angle = alpha * steer_angle + (1-alpha) * prev_steer;
        prev_steer = steer_angle;
        return steer_angle;
    }

    void lidar_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) 
    {   
        // Process each LiDAR scan as per the Follow Gap algorithm & publish an AckermannDriveStamped Message
        std::cout <<  "@=> Start of new scan data" << std::endl;
        // 1. update lidar params
        range_length =    scan_msg->ranges.size();
        range_min =       scan_msg->range_min;
        range_max =       scan_msg->range_max;
        angle_min =       scan_msg->angle_min;
        angle_max =       scan_msg->angle_max;
        angle_increment = scan_msg->angle_increment;
        front_idx_start = angle_to_index(-M_PI / 2);
        front_idx_end   = range_length - front_idx_start; // range(front_idx_start, front_idx_end)
        std::cout << "range length " << range_length << "angle min " << angle_min <<std::endl; 
        std::cout <<  M_PI << "-90~90 indices " << front_idx_start << ", "<< 
        front_idx_end << ", " << scan_msg->angle_min + scan_msg->angle_increment * front_idx_start<<std::endl;
        
        // 2. Preprocess ranges data
        float ranges[range_length];
        preprocess_lidar(&scan_msg->ranges[0], ranges);
        for (int i=0; i<range_length; i++) {
            if (i % 10 == 0) {
                std::cout << std::endl;
            }
            std::cout << ranges[i] << " ";
            
        }
        std::cout << std::endl;

        // 3. Find closest point and (disparity) points and create bubble.
        std::vector<std::pair<int, int>> bubble_indices = find_bubbles(ranges);
        create_bubbles(ranges, bubble_indices);
        std::cout << "BUBBLE_INDEX: " << std::endl;
        for (int i=0; i < bubble_indices.size(); i++) {
            std::cout << bubble_indices[i].first << ", " << bubble_indices[i].second << ";";
        }
        for (int i=0; i < range_length; i++) {
            if (i % 10 == 0) {
                std::cout << std::endl;
            }
            std::cout << ranges[i] << " ";
            
        }

        // 4. Find max length gap
        int gap_indices[2];
        find_max_gap(ranges, gap_indices);
        std::cout <<  std::endl << "gap_indices " << gap_indices[0] << ", stop " << gap_indices[1] << std::endl;
        
        // 5. Find the best point in the gap
        int target_index;
        double angle;
        target_index = find_best_point(ranges, gap_indices);
        angle = scan_msg->angle_min + scan_msg->angle_increment * target_index;
        std::cout <<  std::endl << "best idx " << target_index << ", value " << 
        ranges[target_index] << "heading: " << angle << std::endl;
        
        // 6. Publish Drive message
        ackermann_msgs::msg::AckermannDriveStamped drive_msg;
        drive_msg.drive.steering_angle = get_steering_angle(angle);

        // TODO: SPEED AND STEERING CONTROL
        drive_msg.drive.speed = get_velocity(drive_msg.drive.steering_angle);

        // if (angle > -M_PI/36 or angle < M_PI/36) {
        //     drive_msg.drive.speed = 5.0;
        // }
        // else if ((angle > -M_PI/18 and angle < -M_PI/36) or (angle < M_PI/18 and angle > M_PI/36)) {
        //     drive_msg.drive.speed = 1.0;
        // }
        // else {
        //     drive_msg.drive.speed = 0.5;
        // }
        
        // drive_msg.drive.speed = 1.0;
        publisher_->publish(drive_msg);
    }

};
int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ReactiveFollowGap>());
    rclcpp::shutdown();
    return 0;
}

File: lab-4-follow-the-gap-team-2-1-main/gap_follow/scripts/reactive_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

import numpy as np
from scipy import ndimage
from sensor_msgs.msg import LaserScan
from ackermann_msgs.msg import AckermannDriveStamped

#  Constants from xacro
CHASSIS_WIDTH = 0.2032  # (m)
TIRE_WIDTH = 0.0381  # (m)
CAR_WIDTH = CHASSIS_WIDTH + 2 * TIRE_WIDTH
MAX_STEER = 0.36  # (rad)


class ReactiveFollowGap(Node):
    """
    Implement Wall Following on the car
    This is just a template, you are free to implement your own node!
    """

    def __init__(self):
        super().__init__('reactive_node')

        # Params
        self.declare_parameter('window_size', 5)
        self.declare_parameter('rejection_thres', 2.5)
        self.declare_parameter('distance_thres', 2.0)

        self.declare_parameter('consecutive_hits', 5)
        self.declare_parameter('bubble_radius_factor', 0.7)
        self.declare_parameter('disparity_thres', 0.15)

        self.declare_parameter('k_p', 1.0)
        self.declare_parameter('k_i', 0.0)
        self.declare_parameter('k_d', 0.1)
        self.declare_parameter("alpha", 0.4)

        # self.declare_parameter("low_vel", 0.5)
        # self.declare_parameter("high_vel", 3.0)
        # self.declare_parameter("velocity_attenuation", 1.5)

        self.range_min = None
        self.range_max = None
        self.ranges_length = None       #length of ranges
        self.angle_min = None        # angle range
        self.angle_max = None
        self.angle_increment = None
        self.front_idx_start = None    # idx of scans in -90~90 degrees ahead
        self.front_idx_end = None

        # PID Control Params
        self.prev_angle = 0.0
        self.integral = 0.0
        self.prev_steer = 0.0

        # Topics & Subs, Pubs
        lidarscan_topic = '/scan'
        drive_topic = '/drive'

        self.sub_ = self.create_subscription(LaserScan, lidarscan_topic, self.lidar_callback, 10)
        self.pub_ = self.create_publisher(AckermannDriveStamped, drive_topic, 10)

    def angle_to_index(self, angle: float):
        return int(round((np.clip(angle, self.angle_min, self.angle_max) - self.angle_min) / self.angle_increment))

    def index_to_angle(self, index: int):
        return self.angle_min + index * self.angle_increment

    def preprocess_lidar(self, ranges: np.ndarray):
        window_size = self.get_parameter('window_size').get_parameter_value().integer_value
        rejection_thres = self.get_parameter('rejection_thres').get_parameter_value().double_value
        # Remove invalid readings
        ranges = np.clip(ranges, self.range_min, self.range_max) 
        # Average over window
        ranges = np.convolve(ranges, np.ones(window_size, dtype=float) / window_size, 'same')
        # Clip high values
        ranges[[0, -1]] *= 1.67
        ranges[[1, -2]] *= 1.25
        ranges = np.clip(ranges, 0, rejection_thres)
        return ranges

    def find_disparities(self, ranges):
        disparity_thres = self.get_parameter('disparity_thres').get_parameter_value().double_value
        disparities = np.diff(ranges) # [i] = [i+1] - [i], if 
        left_indices = np.where(disparities > disparity_thres)[0] # bubble spread right
        right_indices = np.where(disparities < -disparity_thres)[0] + 1
        # self.get_logger().info("Num disparities: %0.2f" % len(indices))
        return left_indices, right_indices

    def create_bubbles(self, ranges: np.ndarray):
        bubble_radius = self.get_parameter('bubble_radius_factor').get_parameter_value().double_value * CAR_WIDTH
        closest_indices = [np.argmin(ranges)]
        left_indices, right_indices = self.find_disparities(ranges)
        # indices = np.hstack((closest_indices, disparity_indices))
        self.get_logger().info("Closest {}, left {}, right {}".format(closest_indices, left_indices, right_indices))
        new_ranges = ranges.copy()
        
        for center_idx in closest_indices:
            center_angle = self.index_to_angle(center_idx)
            if bubble_radius < ranges[center_idx]:
                theta = abs(np.arcsin(bubble_radius / ranges[center_idx]))
            else:
                theta = np.pi / 6
                self.get_logger().warning("CLOSE TO WALL; Closest")
            bubble_index_start  = self.angle_to_index(center_angle - theta)
            bubble_index_end    = self.angle_to_index(center_angle + theta)
            new_ranges[bubble_index_start:bubble_index_end] = 0.0
            
        for center_idx in left_indices:
            center_angle = self.index_to_angle(center_idx)
            if bubble_radius < ranges[center_idx]:
                theta = abs(np.arcsin(bubble_radius / ranges[center_idx]))
            else:
                theta = np.pi / 6
                self.get_logger().warning("CLOSE TO WALL; dis")
            bubble_index_start  = center_idx
            bubble_index_end    = self.angle_to_index(center_angle + theta)
            new_ranges[bubble_index_start:bubble_index_end] = ranges[center_idx]
        
        for center_idx in right_indices:
            center_angle = self.index_to_angle(center_idx)
            if bubble_radius < ranges[center_idx]:
                theta = abs(np.arcsin(bubble_radius / ranges[center_idx]))
            else:
                theta = np.pi / 6
                self.get_logger().warning("CLOSE TO WALL; dis")
            bubble_index_start  = self.angle_to_index(center_angle - theta)
            bubble_index_end    = center_idx
            new_ranges[bubble_index_start:bubble_index_end] = ranges[center_idx]
            
        return new_ranges

    def find_max_gap(self, ranges, limited_range=True):
        consecutive_hits = self.get_parameter("consecutive_hits").get_parameter_value().integer_value
        distance_thres = self.get_parameter("distance_thres").get_parameter_value().double_value
        if limited_range:
            ranges = ranges[self.front_idx_start:self.front_idx_end+1]
            offset = self.front_idx_start
        else:
            offset = 0
        extend_ranges = np.hstack((0, ranges, 0))
        zero_indices = np.where(extend_ranges <  distance_thres)[0]
        gap_angles = np.hstack((0, np.diff(zero_indices) - 1))
        right_bound = np.where(gap_angles > 0)[0]
        left_bound = right_bound - 1
        gap_bounds = zero_indices[np.vstack((left_bound, right_bound)).T]
        gap_bounds = gap_bounds + np.array([1, -1])
        gap_bounds = gap_bounds - 1
        print("GAP_BUNDS gap_bounds", gap_bounds)
        gap_lengths = (np.diff(gap_bounds, axis=-1) + 1).flatten()
        
        # max length gap
        # TODO: Max gap selection criteria
        
        gap_bound = gap_bounds[np.argmax(gap_lengths)] + offset
        # print(gap_lengths)
        return gap_bound

    def find_best_point(self, gap_bound, ranges):
        start_i = gap_bound[0]
        end_i = gap_bound[1]
        gap = ranges[start_i:end_i+1]
        max_range = np.max(gap)
        max_ids = np.where(gap == max_range)[0]
        avg_mid = int(np.mean(max_ids))
        return start_i + avg_mid

    def get_steer(self, angle):
        kp = self.get_parameter('k_p').get_parameter_value().double_value
        ki = self.get_parameter('k_i').get_parameter_value().double_value
        kd = self.get_parameter('k_d').get_parameter_value().double_value
        alpha = self.get_parameter('alpha').get_parameter_value().double_value
        d_error = angle - self.prev_angle
        self.prev_angle = angle
        self.integral += angle
        steer = kp * angle + ki * self.integral + kd * d_error
        new_steer = np.clip(steer, -MAX_STEER, MAX_STEER)
        new_steer = alpha * new_steer + (1 - alpha) * self.prev_steer
        self.prev_steer = new_steer
        return new_steer

    # def get_velocity(self, error):
    #     """ Get desired velocity based on current error
    #     """
    #     # Speed is exponential w.r.t error
    #     low_vel = self.get_parameter('low_vel').get_parameter_value().double_value
    #     high_vel = self.get_parameter('high_vel').get_parameter_value().double_value
    #     atten = self.get_parameter('velocity_attenuation').get_parameter_value().double_value

    #     return (high_vel - low_vel) * np.exp(-abs(error) * atten) + low_vel
    def get_velocity(self, angle):
        if abs(angle) < np.pi / 18:
            speed = 1.5
        elif abs(angle) < np.pi / 18 * 2:
            speed = 1.0
        else:
            speed = 0.5
        return speed

    def lidar_callback(self, data: LaserScan):
        ranges = np.array(data.ranges, dtype=float)
        if self.range_min == None:
            self.range_min = data.range_min
            self.range_max = data.range_max
            self.ranges_length = len(ranges)       #length of ranges
            self.angle_min = data.angle_min        # angle range
            self.angle_max = data.angle_max
            self.angle_increment = data.angle_increment
            self.front_idx_start = self.angle_to_index(-np.pi/2)    # idx of scans in -90~90 degrees ahead
            self.front_idx_end = self.ranges_length - self.front_idx_start
        # 2. Preprocess ranges data
        
        # print(list(ranges))
        ranges = self.preprocess_lidar(ranges)
        # print("==================pro")
        # print(list(ranges))
        # 3. create bubble:
        ranges = self.create_bubbles(ranges)
        # print(">>>>>>>>>>>bubbles")
        # print(list(ranges))        
        # Find max length gap
        gap_bound= self.find_max_gap(ranges, limited_range=True)
        self.get_logger().info("Gap: ({}, {})".format(gap_bound[0], gap_bound[1]))

        # Find the best point in the gap
        best_idx = self.find_best_point(gap_bound, ranges)

        self.get_logger().info("best_idx: {}".format(best_idx))

        # Get speed and steer
        angle = self.index_to_angle(best_idx)
        steer = self.get_steer(angle)
        speed = self.get_velocity(steer)
        # steer = self.get_steer(angle)

        # Publish Drive message
        self.get_logger().info("Error: %0.2f,\t Steer: %0.2f,\t Vel: %0.2f" % (np.rad2deg(angle),
                                                                               np.rad2deg(steer),
                                                                               speed))
        message = AckermannDriveStamped()
        message.drive.speed = speed
        message.drive.steering_angle = steer
        self.pub_.publish(message)


def main(args=None):
    rclpy.init(args=args)
    print("WallFollow Initialized")
    reactive_node = ReactiveFollowGap()
    rclpy.spin(reactive_node)

    reactive_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
